app/api/data.py

```diff
@@ -5,43 +5,96 @@
+######################################   Batches      ########################################
+##############################################################################################
+
+
+@api.route('/batches', methods=['GET'])
 def get_batches():
     current_app.logger.info('Retrieving all batches')
     batches = Batch.query.all()
+    # for each batch, attach its coreData rows
+    
     return jsonify({
         'batches': [batch.to_dict() for batch in batches]
     })
 
-@api.route('/data/state/', methods=['GET'])
+
+##############################################################################################
+######################################   States      #########################################
+##############################################################################################
+
+@api.route('/states', methods=['GET'])
 def get_states():
     current_app.logger.info('Retrieving all states')
     states = State.query.all()
     return jsonify({
         'states': [state.to_dict() for state in states]
     })
 
-@api.route('/data/', methods=['GET'])
-def get_core_data():
-    current_app.logger.info('Retrieving all states')
-    data = CoreData.query.join(Batch).join(State).all()
-    return jsonify({
-        'data': [x.to_dict() for x in data]
-    }) 
+##############################################################################################
+######################################   Core data      ######################################
+##############################################################################################
+
 
-@api.route('/data/batch/', methods=['POST'])
-def post_data():
+@api.route('/batches', methods=['POST'])
+def post_core_data():
+    """
+    Workhorse POST method for core data
+
+    Requirements: 
+    """
     current_app.logger.info('Got a post request!')
-    payload = request.json
-    if not payload:
-        abort(400)
-    current_app.logger.info(payload)
-    return jsonify({'payload': payload})
+    payload = request.json  # this is a dict
+
+    # test the input data
+    assert 'context' in payload, "Payload requires 'context' field"
+    assert 'states' in payload, "Payload requires 'states' field"
+    assert 'coreData' in payload, "Payload requires 'coreData' field"
+
+    # we construct the batch from the push context
+    context = payload['context']
+    current_app.logger.info('Creating new batch from context: %s' % context)
+    batch = Batch(**context)
+    db.session.add(batch)
+    db.session.flush()  # this sets the batch ID, which we need for corresponding coreData objects
+
+    # add states
+    state_dicts = payload['states']
+    state_objects = []
+    for state_dict in state_dicts:
+        current_app.logger.info('Creating new state row from info: %s' % state_dict)
+        state = State(**state_dict)
+        state_objects.append(state)
+        db.session.add(state)
+        db.session.flush()
```

Comment:
This works once certainly, but do we need update code for once the state has already been added? I think we'll need to see if the state already exists, and if it does, just update its properties with any new values.